File: implementations/mml_parser.py
"""
MML Parser - Python Implementation
Minimal Markup Language Parser

This is a basic implementation demonstrating core MML parsing functionality.
For production use, consider additional error handling and optimizations.
"""

class MMLParser:
    def __init__(self):
        self.reset()

    def reset(self):
        """Reset parser state"""
        self.document = {
            'type': 'document',
            'title': None,
            'metadata': {},
            'sections': []
        }
        self.current_section = None

    def parse(self, mml_text):
        """
        Parse MML text into document structure

        Args:
            mml_text (str): Raw MML content

        Returns:
            dict: Parsed document object
        """
        self.reset()

        if not mml_text or not isinstance(mml_text, str):
            raise ValueError("Invalid MML input: must be a non-empty string")

        lines = mml_text.split('\n')

        for line in lines:
            line = line.strip()
            if line and ':' in line:
                self._parse_line(line)

        return self.document

    def _parse_line(self, line):
        """Parse a single MML line"""
        colon_index = line.find(':')
        if colon_index == -1:
            return

        tag = line[:colon_index].strip().upper()
        content = line[colon_index + 1:].strip()

        self._process_tag(tag, content)

    def _process_tag(self, tag, content):
        """Process a tag and its content"""
        if tag == 'T':
            self.document['title'] = content

        elif tag == 'H':
            self.current_section = {
                'type': 'section',
                'title': content,
                'metadata': {},
                'children': []
            }
            self.document['sections'].append(self.current_section)

        elif tag == 'P':
            if self.current_section:
                self.current_section['children'].append({
                    'type': 'paragraph',
                    'content': content
                })

        elif tag == 'M':
            parts = content.split('|', 1)
            key = parts[0].strip()
            value = parts[1].strip() if len(parts) > 1 else ''

            metadata_target = self.current_section['metadata'] if self.current_section else self.document['metadata']
            metadata_target[key] = value

        elif tag == 'L':
            parts = content.split('|', 1)
            link_text = parts[0].strip()
            link_url = parts[1].strip() if len(parts) > 1 else ''

            link_node = {
                'type': 'link',
                'text': link_text,
                'url': link_url
            }

            if self.current_section:
                self.current_section['children'].append(link_node)
            else:
                if 'links' not in self.document:
                    self.document['links'] = []
                self.document['links'].append(link_node)

        elif tag == 'Q':
            if self.current_section:
                self.current_section['children'].append({
                    'type': 'quote',
                    'content': content
                })

        elif tag == 'C':
            if self.current_section:
                self.current_section['children'].append({
                    'type': 'code',
                    'content': content
                })

        elif tag == 'IMG':
            parts = content.split('|', 1)
            img_desc = parts[0].strip()
            img_url = parts[1].strip() if len(parts) > 1 else ''

            img_node = {
                'type': 'image',
                'description': img_desc,
                'url': img_url
            }

            if self.current_section:
                self.current_section['children'].append(img_node)

        else:
            # Unknown tag - could log warning
            print(f"Warning: Unknown MML tag: {tag}")

    def to_html(self, doc=None):
        """
        Convert parsed document to HTML

        Args:
            doc (dict, optional): Parsed document. Uses current document if None.

        Returns:
            str: HTML representation
        """
        if doc is None:
            doc = self.document

        html = f"""<!DOCTYPE html>
<html lang="fr">
<head>
    <meta charset="UTF-8">
    <title>{self._escape_html(doc['title'] or 'Document MML')}</title>
</head>
<body>"""

        if doc['title']:
            html += f"<h1>{self._escape_html(doc['title'])}</h1>"

        for section in doc['sections']:
            html += f"<h2>{self._escape_html(section['title'])}</h2>"
            for child in section['children']:
                html += self._node_to_html(child)

        html += "</body></html>"
        return html

    def _node_to_html(self, node):
        """Convert a node to HTML"""
        node_type = node['type']

        if node_type == 'paragraph':
            return f"<p>{self._escape_html(node['content'])}</p>"
        elif node_type == 'quote':
            return f"<blockquote>{self._escape_html(node['content'])}</blockquote>"
        elif node_type == 'code':
            return f"<pre><code>{self._escape_html(node['content'])}</code></pre>"
        elif node_type == 'link':
            return f"<p><a href=\"{self._escape_html(node['url'])}\">{self._escape_html(node['text'])}</a></p>"
        elif node_type == 'image':
            return f"""<figure>
  <img src="{self._escape_html(node['url'])}" alt="{self._escape_html(node['description'])}">
  <figcaption>{self._escape_html(node['description'])}</figcaption>
</figure>"""
        else:
            return f"<p>{self._escape_html(node.get('content', ''))}</p>"

    def _escape_html(self, text):
        """Escape HTML special characters"""
        return (text
                .replace('&', '&amp;')
                .replace('<', '&lt;')
                .replace('>', '&gt;')
                .replace('"', '&quot;')
                .replace("'", '&#39;'))

    def to_json(self, doc=None):
        """
        Convert parsed document to JSON

        Args:
            doc (dict, optional): Parsed document. Uses current document if None.

        Returns:
            str: JSON representation
        """
        import json
        if doc is None:
            doc = self.document
        return json.dumps(doc, indent=2, ensure_ascii=False)


# Example usage
if __name__ == "__main__":
    parser = MMLParser()

    mml_content = """T:Test Document
H:Section 1
P:This is a paragraph
M:Author|John Doe
L:Documentation|https://example.com
Q:Important note
"""

    # Parse the document
    doc = parser.parse(mml_content)
    print("Parsed document:")
    print(parser.to_json(doc))

    print("\nHTML output:")
    print(parser.to_html(doc))

File: implementations/mml-parser.js
/**
 * MML Parser - JavaScript Implementation
 * Minimal Markup Language Parser
 *
 * This is a basic implementation demonstrating core MML parsing functionality.
 * For production use, consider additional error handling and optimizations.
 */

class MMLParser {
  constructor() {
    this.reset();
  }

  reset() {
    this.document = {
      type: 'document',
      title: null,
      metadata: {},
      sections: []
    };
    this.currentSection = null;
  }

  /**
   * Parse MML text into document structure
   * @param {string} mmlText - Raw MML content
   * @returns {object} Parsed document object
   */
  parse(mmlText) {
    this.reset();

    if (!mmlText || typeof mmlText !== 'string') {
      throw new Error('Invalid MML input: must be a non-empty string');
    }

    const lines = mmlText.split('\n');

    for (const line of lines) {
      const trimmedLine = line.trim();
      if (trimmedLine && trimmedLine.includes(':')) {
        this.parseLine(trimmedLine);
      }
    }

    return this.document;
  }

  /**
   * Parse a single MML line
   * @param {string} line - Single MML line
   */
  parseLine(line) {
    const colonIndex = line.indexOf(':');
    if (colonIndex === -1) return;

    const tag = line.substring(0, colonIndex).trim().toUpperCase();
    const content = line.substring(colonIndex + 1).trim();

    this.processTag(tag, content);
  }

  /**
   * Process a tag and its content
   * @param {string} tag - MML tag
   * @param {string} content - Tag content
   */
  processTag(tag, content) {
    switch (tag) {
      case 'T':
        this.document.title = content;
        break;

      case 'H':
        this.currentSection = {
          type: 'section',
          title: content,
          metadata: {},
          children: []
        };
        this.document.sections.push(this.currentSection);
        break;

      case 'P':
        if (this.currentSection) {
          this.currentSection.children.push({
            type: 'paragraph',
            content: content
          });
        }
        break;

      case 'M':
        const [key, value] = content.split('|', 2);
        const metadataTarget = this.currentSection ?
          this.currentSection.metadata :
          this.document.metadata;

        metadataTarget[key.trim()] = value ? value.trim() : '';
        break;

      case 'L':
        const [linkText, linkUrl] = content.split('|', 2);
        const linkNode = {
          type: 'link',
          text: linkText.trim(),
          url: linkUrl ? linkUrl.trim() : ''
        };

        if (this.currentSection) {
          this.currentSection.children.push(linkNode);
        } else {
          if (!this.document.links) this.document.links = [];
          this.document.links.push(linkNode);
        }
        break;

      case 'Q':
        if (this.currentSection) {
          this.currentSection.children.push({
            type: 'quote',
            content: content
          });
        }
        break;

      case 'C':
        if (this.currentSection) {
          this.currentSection.children.push({
            type: 'code',
            content: content
          });
        }
        break;

      case 'IMG':
        const [imgDesc, imgUrl] = content.split('|', 2);
        const imgNode = {
          type: 'image',
          description: imgDesc.trim(),
          url: imgUrl ? imgUrl.trim() : ''
        };

        if (this.currentSection) {
          this.currentSection.children.push(imgNode);
        }
        break;

      // Add more tags as needed...

      default:
        // Unknown tag - could log warning or handle gracefully
        console.warn(`Unknown MML tag: ${tag}`);
    }
  }

  /**
   * Convert parsed document to HTML
   * @param {object} doc - Parsed MML document
   * @returns {string} HTML representation
   */
  toHTML(doc = this.document) {
    let html = `<!DOCTYPE html>
<html lang="fr">
<head>
    <meta charset="UTF-8">
    <title>${this.escapeHTML(doc.title || 'Document MML')}</title>
</head>
<body>`;

    if (doc.title) {
      html += `<h1>${this.escapeHTML(doc.title)}</h1>`;
    }

    for (const section of doc.sections) {
      html += `<h2>${this.escapeHTML(section.title)}</h2>`;
      for (const child of section.children) {
        html += this.nodeToHTML(child);
      }
    }

    html += '</body></html>';
    return html;
  }

  /**
   * Convert a node to HTML
   * @param {object} node - Document node
   * @returns {string} HTML string
   */
  nodeToHTML(node) {
    switch (node.type) {
      case 'paragraph':
        return `<p>${this.escapeHTML(node.content)}</p>`;
      case 'quote':
        return `<blockquote>${this.escapeHTML(node.content)}</blockquote>`;
      case 'code':
        return `<pre><code>${this.escapeHTML(node.content)}</code></pre>`;
      case 'link':
        return `<p><a href="${this.escapeHTML(node.url)}">${this.escapeHTML(node.text)}</a></p>`;
      case 'image':
        return `<figure>
  <img src="${this.escapeHTML(node.url)}" alt="${this.escapeHTML(node.description)}">
  <figcaption>${this.escapeHTML(node.description)}</figcaption>
</figure>`;
      default:
        return `<p>${this.escapeHTML(node.content || '')}</p>`;
    }
  }

  /**
   * Escape HTML special characters
   * @param {string} text - Text to escape
   * @returns {string} Escaped text
   */
  escapeHTML(text) {
    return text
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#39;');
  }

  /**
   * Convert parsed document to JSON
   * @param {object} doc - Parsed MML document
   * @returns {string} JSON representation
   */
  toJSON(doc = this.document) {
    return JSON.stringify(doc, null, 2);
  }
}

// Export for use in Node.js or browsers
if (typeof module !== 'undefined' && module.exports) {
  module.exports = MMLParser;
}

// Example usage:
/*
const parser = new MMLParser();
const doc = parser.parse(`
T:Test Document
H:Section 1
P:This is a paragraph
M:Author|John Doe
`);

console.log(doc);
console.log(parser.toHTML(doc));
*/
